### ISSUE TITLE
Registry addon cannot be accessed internally when using cri-o runtime

### ISSUE THREAD

**Issue by tliron** [kubernetes/minikube#6982]
**The exact command to reproduce the issue**:
```
minikube start --addons=registry --network-plugin=cni --enable-default-cni
```

**The full output of the command that failed**:

Pods fail to pull image from the registry, e.g.:
```
image: localhost:5000/mycontainer
```

Note that I *can* push to the registry from outside via `localhost:5000/mycontainer`.

Switching to Docker runtime allows this to work.

**The operating system version**:

Fedora 31

**Comment by afbjorklund** [comment]
You are saying that the registry works with `--container-runtime=docker`, but not with `--container-runtime=cri-o` ? (the above command-line is more talking about CNI options, rather than CRI...)

From the outside, it is normally more like: https://minikube.sigs.k8s.io/docs/tasks/docker_registry/

**Comment by tliron** [comment]
Ugh, so sorry, did a wrong copy and paste. Will edit the original issue,

**Comment by afbjorklund** [comment]
Do you have the error output, like if you do a `describe pod` on it ?

**Comment by tliron** [comment]
These are the events:
```
16s         Normal    Pulling             pod/mycontainer-f7dcbd7f6-tx9nc    Pulling image "localhost:5000/mycontainer"
28s         Warning   Failed              pod/mycontainer-f7dcbd7f6-tx9nc    Failed to pull image "localhost:5000/mycontainer": rpc error: code = Unknown desc = error pinging docker registry localhost:5000: Get http://localhost:5000/v2/: dial tcp 127.0.0.1:5000: i/o timeout
```
Nothing much more in describe, just the fail reason being `ImagePullBackOff`.

**Comment by afbjorklund** [comment]
There doesn't seem to be anything in the registry logs, so I guess it is getting stuck in registry-proxy

It works from the outside, but just hangs from the inside. Even with `curl`, so something is wrong...


The internal IP are fine:
```
$ curl http://10.1.0.2:5000/v2/
{}$ 
$ curl http://10.1.0.3:80/v2/
{}$ 
```

But not the localhost.

```
$ curl http://192.168.99.100:5000/v2/
{}$ 
$ curl http://localhost:5000/v2/
curl: (7) Failed to connect to localhost port 5000: Connection timed out

```

**Comment by afbjorklund** [comment]
Looks similar to https://github.com/cri-o/cri-o/issues/1804

**Comment by tliron** [comment]
Thank you for verifying. Any ideas of a temporary workaround?

**Comment by afbjorklund** [comment]
Note that it _has_ to be localhost, since that is the hack used to get an insecure registry...

`Failed to pull image "192.168.99.100:5000/test-img": rpc error: code = Unknown desc = error pinging docker registry 192.168.99.100:5000: Get https://192.168.99.100:5000/v2/: http: server gave HTTP response to HTTPS client`

The workaround is configuring the minikube IP as an insecure registry, in the crio.conf

```
# List of registries to skip TLS verification for pulling images. Please   
# consider configuring the registries via /etc/containers/registries.conf before
# changing them here.                                                          
#insecure_registries = "[]"                                                      
```

---

Or deploy a proper https registry (left as an exercise for the reader :-) )

https://docs.docker.com/registry/deploying/#run-an-externally-accessible-registry

**Comment by tliron** [comment]
Hm, so it looks like cri-o doesn't support Docker's "localhost is allowed as an insecure registry" hack? If that's the case, I think it should be documented that the registry addon doesn't work with cri-o, at least this time.

**Comment by afbjorklund** [comment]
Yeah, it seems like hostPort has some issues with cri-o and that should probably be documented...

**Comment by afbjorklund** [comment]
> Hm, so it looks like cri-o doesn't support Docker's "localhost is allowed as an insecure registry" hack?

That part is supported just fine, since it tries to be bug-compatible at least for the docker:// protocol

**Comment by tliron** [comment]
I found a simple workaround. Get the registry service's cluster IP:
```
kubectl get svc registry --output=jsonpath='{.spec.clusterIP}' --namespace=kube-system
```
And then use that address, with port 80, for the image URL, e.g.:
```
image: 10.102.27.206:80/mycontainer
```

**Comment by priyawadhwa** [comment]
Hey @tliron awesome. We should update the documentation to include that workaround when trying to use the registry addon with cri-o.

**Comment by tliron** [comment]
@priyawadhwa Thanks, but it's an awkward solution. The problem is that users would have to update all their specs to use this varying IP address. I hope we can find a way to enable `localhost:5000` for CRI-O (which, we hope, will become minikube's default runtime).

**Comment by tliron** [comment]
Here is a version of my workaround above in Go code:
```go
import (
	"context"
	"fmt"
	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes" // v0.18.0
)

func GetInternalRegistryURL(kubernetesClient *kubernetes.Clientset) (string, error) {
	if service, err := kubernetesClient.CoreV1().Services("kube-system").Get(context.TODO(), "registry", meta.GetOptions{}); err == nil {
		return fmt.Sprintf("%s:80", service.Spec.ClusterIP), nil
	} else {
		return "", err
	}
}
```

**Comment by charandas** [comment]
The same thing happens with ingress addon. No go with cri-o runtime. The minikube ip does not host any loadbalancer, and it fails silently, while internally, the `ingress` objects bind to minikube ip just fine.

**Comment by medyagh** [comment]
I would be happy to review any PR that fixes this for CRI-O

**Comment by CtrlZvi** [comment]
/remove-lifecycle rotten
To the best of my knowledge, this is still a valid issue.

**Comment by tliron** [comment]
I think an acceptable solution is to document this in Minikube, probably [here](https://minikube.sigs.k8s.io/docs/handbook/registry/). To be honest, that whole page could use a rewrite, as it's more a collection of notes than a guide. As it stands right now, the only way to really learn what `--addons=registry` does is to look at blogs.

I'd like to help, but I'm unfamiliar with the project interactions. Could someone who knows please take responsibility for resolving this issue?

**Comment by medyagh** [comment]
> I think an acceptable solution is to document this in Minikube, probably [here](https://minikube.sigs.k8s.io/docs/handbook/registry/). To be honest, that whole page could use a rewrite, as it's more a collection of notes than a guide. As it stands right now, the only way to really learn what `--addons=registry` does is to look at blogs.
> 
> I'd like to help, but I'm unfamiliar with the project interactions. Could someone who knows please take responsibility for resolving this issue?

I agree with you @tliron  that page is getting very messy and hard to read, I would accept a PR that organizes that page on our webiste and makes it easier to read

**Comment by tliron** [comment]
It sounds like I'm being volunteered. :) I may try to take a stab at it if I can find the time.

It's all very hard, to be honest. The challenge is not just to allow Kubernetes's container runtime access to the private registry (with authentication? authorization?) but also how a user/developer can access it push images to it. After all, that is the goal of having such a built-in registry in the first place. None of this work is trivial. But at least it can be documented specifically for Minikube, which does have some advantages in that it can run in "insecure" mode (self-signed cert).

On that note, I ended up creating a generic solution to the problem of handling and working with private registries, a project I call [Reposure](https://github.com/tliron/reposure). Reposure has explicit support for Minikube's registry add-on.

**Comment by medyagh** [comment]
> It sounds like I'm being volunteered. :) I may try to take a stab at it if I can find the time.
> 
> It's all very hard, to be honest. The challenge is not just to allow Kubernetes's container runtime access to the private registry (with authentication? authorization?) but also how a user/developer can access it push images to it. After all, that is the goal of having such a built-in registry in the first place. None of this work is trivial. But at least it can be documented specifically for Minikube, which does have some advantages in that it can run in "insecure" mode (self-signed cert).
> 
> On that note, I ended up creating a generic solution to the problem of handling and working with private registries, a project I call [Reposure](https://github.com/tliron/reposure). Reposure has explicit support for Minikube's registry add-on.

@tliron thank you That does sounds like a good idea to update our docs and maybe refactor our docs on registry addon with better examples and better categorization

**Comment by iacker** [comment]
I am working on a docs PR to clarify registry addon behavior and a practical workaround for CRI-O users, based on the guidance already captured in this issue. /assign
